### PR TITLE
[dynamo] Fix incorrectly casting `SymNode` to `int` when input is `bool`

### LINKED ISSUE #113393: torch.compile() results in inflexible model with mistralai/Mistral-7B-v0.1
### 🐛 Describe the bug

When applying torch.compile() to HF model `mistralai/Mistral-7B-v0.1`, the resulting model is inflexible in sequence length. The repro code and error message is below:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained(
            "mistralai/Mistral-7B-v0.1",
            torch_dtype=torch.float16,,
            use_cache=True,
        ).to(torch.device('cuda'))

model = torch.compile(model)

#Below three lines work as expected
input_ids =torch.randint(low=0, high=60000, size=(1, 32), dtype=torch.int64)
attention_mask = torch.ones(1, 32, dtype=torch.int64)
model(input_ids=input_ids, attention_mask=attention_mask)

#Next three lines throw error
input_ids =torch.randint(low=0, high=60000, size=(1, 64), dtype=torch.int64)
attention_mask = torch.ones(1, 64, dtype=torch.int64)
model(input_ids=input_ids, attention_mask=attention_mask)

'''
torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in function not_>(*(Eq(s1, s0) & Eq(s2, s0),), **{}):
The expression should not be static by this point: False
 
from user code:
   File "/dev_data/petermca/python_packages/transformers/models/mistral/modeling_mistral.py", line 295, in forward
    if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
 
Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
 
You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True"
'''
```

### Versions

torch==2.2.0.dev20231108+cu118
transformers==4.34.1

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/113393, https://github.com/pytorch/pytorch/pull/113848#issuecomment-1814624510

Incorrectly casting symnode type will cause it to take the wrong path in symbolic_shapes

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1490,6 +1490,10 @@ def _unimplemented():
 
     # and_ is a constant fold function, so we only get here if constant fold is not valid
     def call_and_(self, tx, a, b):
+        if isinstance(a, ConstantVariable) and isinstance(b, ConstantVariable):
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builtin.py:1493]: There is actually something fishy going on. Why is constant_handler not being called? Should we not use constant_handler before trying to evaluate handler? I find this strange.

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1493]: I've changed the behaviour to fallback on constant handler. Although I still find it odd. Why is constant_handler the fallback? Why don't we try the constant handler first?? :thinking:

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1493]: I'm gonna do a refactor where we get rid of all the constant variable special-casing in individual handlers due to this poor ordering. Heads up @jansel @ezyang @lezcano

**Review by lezcano — COMMENTED** [review]
Are we using `operator.and_` as a synonym for `and`? If so, that's not correct

**Comment by jon-chuang** [comment]
> Are we using operator.and_ as a synonym for and? If so, that's not correct

Actually, `operator.and_` behaves the same way as `and`.

```python
operator.and_(True, 1)  # 1
```

**Comment by jon-chuang** [comment]
Anw @lezcano that's unrelated to this PR. The same issue would have occured even if we used builtin and semantics.

This discussion more relevant to https://github.com/pytorch/pytorch/pull/113848#issuecomment-1814643569.

**Comment by jon-chuang** [comment]
Tests failing due to this: https://github.com/pytorch/pytorch/issues/113875

**Review by jansel — APPROVED** [review]
I think it would be better to do the refactor, but this change seems better than nothing.

**Comment by jon-chuang** [comment]
@jansel this will now become a special case of https://github.com/pytorch/pytorch/pull/113893, with additional modification to never try to cast `SymNodeVariable` sym_value to `int` from `bool`, which was the root cause.

**Pull request closed without merging**